src/execution_plan/ops/shared/update_functions.c

```diff
@@ -180,25 +176,44 @@ void EvalEntityUpdates
 		update = rm_malloc(sizeof(PendingUpdateCtx));
 		update->ge            = entity;
 		update->attributes    = AttributeSet_ShallowClone(*entity->attributes);
-		update->add_labels    = NULL;
-		update->remove_labels = NULL;
 		// add update context to updates dictionary
 		HashTableAdd(updates, (void *)ENTITY_GET_ID(entity), update);
 	} else {
 		// update context already exists
 		update = (PendingUpdateCtx *)HashTableGetVal(entry);
 	}
 
-	if(array_len(ctx->add_labels) > 0 && update->add_labels == NULL) {
-		update->add_labels = array_new(const char *, array_len(ctx->add_labels));
+	if(array_len(ctx->add_labels) > 0 && *add_labels == NULL) {
+		RG_Matrix_new(add_labels, GrB_BOOL, Graph_RequiredMatrixDim(gc->g),
+			Graph_RequiredMatrixDim(gc->g));
 	}
 
-	if(array_len(ctx->remove_labels) > 0 && update->remove_labels == NULL) {
-		update->remove_labels = array_new(const char *, array_len(ctx->remove_labels));
+	if(array_len(ctx->remove_labels) > 0 && *remove_labels == NULL) {
+		RG_Matrix_new(remove_labels, GrB_BOOL, Graph_RequiredMatrixDim(gc->g),
+			Graph_RequiredMatrixDim(gc->g));
 	}
 	
-	array_union(update->add_labels, ctx->add_labels, strcmp);
-	array_union(update->remove_labels, ctx->remove_labels, strcmp);
+	for (uint i = 0; i < array_len(ctx->add_labels); i++) {
+		const char *label = ctx->add_labels[i];
+		const Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_NODE);
+		if(s == NULL) {
+			s = AddSchema(gc, label, SCHEMA_NODE, true);
+		}
+
+		int schema_id = Schema_GetID(s);
+		RG_Matrix_setElement_BOOL(*add_labels, ENTITY_GET_ID(entity), schema_id);
+	}
+
+	for (uint i = 0; i < array_len(ctx->remove_labels); i++) {
+		const char *label = ctx->remove_labels[i];
+		const Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_NODE);
+		if(s == NULL) {
+			s = AddSchema(gc, label, SCHEMA_NODE, true);
+		}
```

Comment:
Do NOT create a scheme in this case, consider
`MATCH (n:N) REMOVE n:NoneExistingSchema`
There's no harm "removing" a non existing schema from a node. obviously the node isn't labeled under such schema...

Simply `continue`